[CIR] floating-point, pointer, and function types

dkolsen-pgi opened the pull request:
Upstream ClangIR support for `void` type, floating-point types, pointer types, and function types.

Floating-point support is missing the IBM double-double format, because that hasn't been implemented in the incubator project yet.

Pointer types do not yet support address spaces.

Function type support includes only the return type and the parameter types.  The many other properties and attributes of function types will be upstreamed later.

dkolsen-pgi committed:
[CIR] floating-point, pointer, and function types

Upstream ClangIR support for `void` type, floating-point types, pointer
types, and function types.

Floating-point support is missing the IBM double-double format, because
that hasn't been implemented in the incubator project yet.

Pointer types do not yet support address spaces.

Function type support includes only the return type and the parameter
types.  The many other properties and attributes of function types will
be upstreamed later.

efriedma-quic commented on clang/include/clang/CIR/Dialect/IR/CIRTypes.td:
```diff
@@ -129,4 +130,224 @@ def PrimitiveInt
+    A floating-point type that represents the `long double` type in C and C++.
+
+    The underlying floating-point format of a `long double` value depends on the
+    target platform and the implementation. The `underlying` parameter specifies
```
FYI, clang supports targets where "double" is 32 bits.  Granted, we're talking about pretty obscure targets, but maybe worth implementing now while you're looking at the relevant code.

dkolsen-pgi replied:
We'll have to tackle the more unusual and lesser-used targets, such as ones with 32-bit `double`, when we get there.  I can't get that code correct without having such a target to test on.

This effort is primarily upstreaming existing code from the ClangIR incubator project (which currently has limited target support) to the main LLVM repository.  I prefer to limit the amount of new code and new functionality as part of the upstreaming process.

erichkeane replied:
I wonder if we should have a 'not yet implemented' diagnostic that checks the size of 'double' to make sure it is 8?  That way we don't end up generating invalid code.  (Basically, check `ASTContext::getTypeSizeInChars` during setup?).

dkolsen-pgi replied:
Good idea.  But I'm not sure where in the setup code to put such a check, since I need to check the format of the Clang types `float` and `double` and I'm not sure how best to access them in the ClangIR setup code.  Instead I have added `assert`s to the `BuiltinType::Float` and `BuiltinType::Double` cases in `CIRGenTypes::convertType`.  The `assert` will fire if the source code tries to use `float` or `double` on a platform where they aren't IEEE 32-bit and 64-bit.  The type `long double` already has a similar check in `CIRGenBuilderTy::getLongDoubleTy`.  I believe all the other non-standard floating-point types have fixed formats that shouldn't depend on the target, so they don't need similar checks.

erichkeane replied:
I think those asserts are good enough, thanks!

bcardosolopes replied:
Those asserts are probably the best way to go for now.

erichkeane commented on clang/include/clang/CIR/Interfaces/CIRFPTypeInterface.td:
```diff
@@ -0,0 +1,52 @@
+//===- CIRFPTypeInterface.td - CIR FP Interface Definitions -----*- C++ -*-===//
```
Wasn't there some discussion on the last patch about the header not containing certain info?  I see this has the top level header, and the above file (CIRFPTypeInterface.h) having the description.  Which were we supposed to skip?

dkolsen-pgi replied:
You're right.  The updated guidelines say that this line doesn't need to contain any useful information.  I'll remove it.

erichkeane replied:
If you would, please audit the rest of the files that get added, I thought I noticed it a few places scrolling through.

dkolsen-pgi replied:
Done.  `CIRFPTypeInterface.cpp` had the same issue, which I fixed in a soon-to-be-commited change.

I don't plan to make this change in any existing files that were added in earlier commits, only in new files going forward.

erichkeane replied:
Yep, makes sense to me.

erichkeane commented on clang/lib/CIR/CodeGen/CIRGenTypes.cpp:
```diff
@@ -18,6 +21,87 @@ mlir::MLIRContext &CIRGenTypes::getMLIRContext() const {
+  // certain circumstances have been reached.
+  assert(!type->getAs<MemberPointerType>() && "NYI");
+
+  // If this isn't a tagged type, we can convert it!
```
First, `tag` type, not `tagged` type, as this just represents any of the struct, class, union, or enum types.

Second, is the rest of this true?  Non-Tag types include atomics, Attributed types, bitint types, block pointers, complex, various dependent types, vector types, etc.  Do we really just mean to do `BuiltinType` here instead?  Or do we really mean we can do the rest already?

dkolsen-pgi replied:
This function is used to check if a type is problematic when it is used as part of a function type.  According to the comment here, only incomplete `class`/`struct`/`union` types are problematic when creating ClangIR function types.

The currently upstreamed ClangIR code can't handle atomic types or vector types or some of the other types you mentioned, in any context.  But when ClangIR can handle them, they will just work as part of function types; the code here won't need to change.

This entire function could have been left out of this PR since it doesn't do anything useful yet.  But because everything in the function except for the last line builds and runs, I decided to include it in this PR.

erichkeane replied:
In that case, I think just an assert in the  '!TagType' check here would be valuable, just that asserts it is a `BuiltinType` (or, the NYI diagnostic).

dkolsen-pgi replied:
I disagree about this one.  If the source code uses a vector type as the type of a function parameter, ClangIR code gen will fail with a NYI error in `CIRGenTypes::convertType` when trying to convert the Clang vector type into a CIR type.  There isn't a good reason to add a general-purpose check here in a function that is looking for a very specific condition that has only to do with function types.

Maybe this function should be renamed to `isConvertedTypeUsableInFunctionTypes`, to emphasize that it is not checking if a type is convertible, but checking whether or not a type is problematic if used as a return type or a parameter type within a function type.

erichkeane replied:
IMO, the assert here makes a lot of sense, in that it shows off the assumption that we're intentionally not handling these types.  The point is that we're NOT doing the work to figure out if those types are problematic if used in return /parameter in a function type (for good reason, since it'll fail earlier).  SO we should assert for that, so that when `CIRGenTypes::convertType` stops emitting the NYI, that this will be quickly noticed, rather than returning a value we wouldn't be sure of.

erichkeane replied:
>It is known that incomplete tag types are the only ones that are problematic in this context.

This seems inaccurate to me.  CIR doesn't seem to be able to represent(see here: https://github.com/llvm/clangir/blob/main/clang/include/clang/CIR/Dialect/IR/CIRTypes.td)  Atomic Types, BitInt types, Block types, and Matrix types upon quick look (that is perhaps an incomplete list).  

The function was written across; 
https://github.com/llvm/clangir/commit/101e732565d0e5004c5191b0053eebd8e849cade
and 
https://github.com/llvm/clangir/commit/93f55f6d43be394250009013c83d343be7c63f41

Neither of which mention those types specifically?  It seems it USED to have an assert for not-tag-type, though it isn't clear that the authors considered the above types.

Can @lanza and @bcardosolopes chime in and let me know if I'm misinterpreting the commits there?

bcardosolopes replied:
I think it's totally fine to be conservative about assumptions, we need asserts for both things that are NYI but also for things we assume as invariants, IMO the more the merrier cause it's easier to catch miscompilations.

> Neither of which mention those types specifically? It seems it USED to have an assert for not-tag-type, though it isn't clear that the authors considered the above types.

I don't remember the full context anymore, but assertion on a assumed invariant in the incubator to be even more protective also sounds good to me (guess running the tests upon such change will give us the final answer).

dkolsen-pgi replied:
@bcardosolopes What exactly do you suggest doing here in this function.  Asserting on assumed invariants is great.  But I don't see any invariants here that could be asserted.

bcardosolopes replied:
I don't remember the details on why it was necessary at the time to deviate - so I just made this be equivalent to traditional codegen and everything passes: https://github.com/llvm/clangir/commit/04d7dcfb2582753f3eccbf01ec900d60297cbf4b

That said, I don't think we need to introduce any asserts here, can't get better than OG for the sake of completeness. Not supported types will crash given the other mechanisms David mentioned, but the recursive logic is sound with pre-existing work.

Unless Erich has any other concerns, this LGTM to land.

erichkeane replied:
If this includes llvm/clangir@04d7dcf in it, rather than the current implementation, I think I'm OK without any asserts.  I'll Approve, but conditional on that patch being included in this one.

erichkeane commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -133,6 +143,276 @@ IntType::verify(llvm::function_ref<mlir::InFlightDiagnostic()> emitError,
+      p << ", ";
+    p << "...";
+  }
+  p << ')';
```
Why the close paren only in var-args?  What about in all other cases?

dkolsen-pgi replied:
Printing the ')' is not in the varargs section.  It is at the top level of the function.  It is always output.

erichkeane replied:
Coulda sworn I triple checked that before commenting.. Sorry about that.

dkolsen-pgi committed:
[CIR] Follow-up to floating-point upstream change

Fix the opening comment on some newly added files.

Add asserts that the Clang types `float` and `double` are IEEE 32-bit
and 64-bit.

keryell commented:
@dkolsen-pgi  Great PR!
Since it touches functions returning `!void` or `!cir.void`, I made some changes recently with https://github.com/llvm/clangir/commit/568b51537e573c7a8fb616cda5cdd0aa54bc0832 to avoid disrupting some MLIR invariants.
Are you up-streaming the changes in commit order or are you up-streaming the changes with the latest version of a feature?
@joker-eph How to up-stream some changes which are cross-project like https://github.com/llvm/clangir/pull/1203 which introduces a new MLIR parser/pretty-printer feature like:
https://github.com/llvm/clangir/pull/1203/files#diff-cfea91beb87a3b28295e5612974f0556b4daf9a42c1adeb65e56f0dec062feaa
https://github.com/llvm/clangir/pull/1203/files#diff-6c1c134d749d53cf770548806b955401e47b1f15ffe084905a0b132d73b0b1fe
https://github.com/llvm/clangir/pull/1203/files#diff-dabff257abfd11fded480c2dfc0c2afdf640769384af9e37820e11003a75edb1
to allow some keywords starting with `!`?

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenTypes.cpp:
```diff
@@ -63,13 +153,71 @@ mlir::Type CIRGenTypes::convertType(QualType type) {
+      resultType = cgm.FloatTy;
+      break;
+    case BuiltinType::Double:
+      assert(&astContext.getFloatTypeSemantics(type) ==
```
If we know there are possible 32 bit doubles (given previous comment), this looks more like NYI?

dkolsen-pgi replied:
I am adding "NYI" to the assert messages.  But I like keeping these as asserts rather than compiler errors because `float` and `double` with unusual formats is not something we will support anytime soon, so the more drastic assertion failure seems more appropriate.

bcardosolopes commented on clang/lib/CIR/Dialect/IR/CIRTypes.cpp:
```diff
@@ -133,6 +143,276 @@ IntType::verify(llvm::function_ref<mlir::InFlightDiagnostic()> emitError,
+llvm::TypeSize
+PointerType::getTypeSizeInBits(const ::mlir::DataLayout &dataLayout,
+                               ::mlir::DataLayoutEntryListRef params) const {
+  // FIXME: improve this in face of address spaces
```
Can we get MissingFeature asserts for the ones here or below? Might be worth introducing that upstream if we haven't already?

dkolsen-pgi replied:
You are right that a MissingFeature assert would be appropriate here.  The MissingFeature asserts are not in upstream yet.  I haven't felt an urge to migrate them yet because I find them confusing, and have to look at the code a couple times to understand what is happening.  In a (hopefully not-too-distant) future commit I will look at the MissingFeature asserts in the incubator and put something similar in upstream.  But not in this commit.

bcardosolopes reviewed (approved):
LGTM with some minor nits

bcardosolopes commented:
@keryell 

> How to up-stream some changes which are cross-project like https://github.com/llvm/clangir/pull/1203 which introduces a new MLIR parser/pretty-printer feature like:

My bad here, we usually don't accept such changes and I missed that during review time. ClangIR policy to changes outside CIR is to upstream them to MLIR and after that's done we cherry-pick it to the incubator (so rebases are smooth). I'll have to revert your change, sorry for the churn!

keryell commented:
> My bad here, we usually don't accept such changes and I missed that during review time. ClangIR policy to changes outside CIR is to upstream them to MLIR and after that's done we cherry-pick it to the incubator (so rebases are smooth). I'll have to revert your change, sorry for the churn!

@bcardosolopes I can work on the phase 2 of the change I had in mind and remove also the `!void` from function type textual IR to avoid adding a new feature in the MLIR parsing infrastructure.

bcardosolopes commented:
> @bcardosolopes I can work on the phase 2 of the change I had in mind and remove also the `!void` from function type textual IR to avoid adding a new feature in the MLIR parsing infrastructure.

Neat, if you can do that I won't need to revert and make it easier for everyone

dkolsen-pgi committed:
[CIR] Improve a couple assert messages

dkolsen-pgi commented:
@keryell :
> Are you up-streaming the changes in commit order or are you up-streaming the changes with the latest version of a feature?

Upstreaming is happening in logical chunks, using the latest version of the feature.  I am paying no attention to the order of commits in the incubator project.  Unfortunately, your change landed after I started working on this change (or at least after I had last refreshed my incubator workspace).  So I didn't see it, and didn't incorporate your change into this one.

Once this PR is merged, could you please make your change again in the LLVM repository?  Bruno floated this idea in #clangir on Discord, of asking people who change already upstreamed code to make the change in both the incubator and the LLVM repository.  You may be the first guinea pig for this.  We can see if this is a good process.

erichkeane reviewed (approved):
Approve, conditional on isFuncParamTypeConvertible change being pulled from the incubator.

dkolsen-pgi committed:
[CIR] Merge change to CIRGenTypes::isFuncParamTypeConvertible

Merge in change https://github.com/llvm/clangir/commit/04d7dcf
which simplifies function CIRGenTypes::isFuncParamTypeConvertible.
